src/odemis/driver/symphotime.py

```diff
@@ -0,0 +1,891 @@
+        self.iNACKRecNumber = iNACKRecNumber
+        self.iOptRecordCount = len(odOptional)
+        self.odOptional = odOptional
+
+    def __str__(self):
+        return '[DATAFRAME_SRVNAC]'
+
+
+class BasicDataFlow(model.DataFlow):
+
+    def __init__(self, start_func, stop_func):
+        """
+        start_func: function to execute when start_generate is called
+        stop_func: function to execute when stop_generate is called
+        """
+        model.DataFlow.__init__(self)
+        self._start = start_func
+        self._stop = stop_func
+
+    # start/stop_generate are _never_ called simultaneously (thread-safe)
+    def start_generate(self):
+        self._start()
+
+    def stop_generate(self):
+        self._stop()
+
+
+class BEvent(object):
+
+    def __init__(self):
+        self._positive = threading.Event()
+        self._negative = threading.Event()
+        # by default, positive is false, so negative is true
+        self._negative.set()
+
+    def set(self):
+        self._positive.set()
+        self._negative.clear()
+
+    def clear(self):
+        self._positive.clear()
+        self._negative.set()
+
+    def is_set(self):
+        return self._positive.is_set()
+
+    def wait_true(self):
+        self._positive.wait()
+
+    def wait_false(self):
+        self._negative.wait()
+
+
+class Controller(model.Detector):
+    '''
+    A Symphotime Server Parent controller, defined as a detector
+
+    Public:
+        data (BasicDataFlow): Detector data flow
+    '''
+
+    def __init__(self, name, role, children, host, port=DEFAULT_PORT, **kwargs):
+        """
+        children (dict str -> dict): internal role -> kwargs. The internal roles
+          can be "scanner"
+        host: the TCP/IP hostname of the server
+        port: the TCP/IP port of the server. 
+
+        Raises:
+            ValueError if no scanner child is present
+        """
+        super(Controller, self).__init__(name, role, **kwargs)
+        
+        if not children:
+            raise ValueError("Symphotime detector requires a scanner child. ")
+        
+        self._host = host
+        self._port = port
+        self._is_connected = False
+
+        try:
+            logging.debug("Connecting to %s:%d", self._host, self._port)
+            self._socket = socket.create_connection((self._host, self._port))
+            self._socket.settimeout(2.0)
+
+        except socket.error:
+            raise model.HwError("Failed to connect to '%s:%d', check that the Symphotime "
+                                "Server is connected to the network, turned "
+                                "on, and correctly configured." % (host, port))
+
+        # to acquire before sending anything on the socket
+        self._net_access = threading.Lock()
+        
+        # try get parameters from metadata
+        self._metadata[model.MD_PIXEL_SIZE] = (10e-6, 10e-6)
+        self._metadata[model.MD_DESCRIPTION] = ""
```

Comment:
No need to set `MD_DESCRIPTION` here, as it's optional. However, I'd suggest to mention in the docstring of the Component the metadata which are handled. (ie, MD_PIXEL_SIZE, MD_DESCRIPTION, MD_LENS_NAME).